### ISSUE TITLE
V18 hangs on main thread when unresolved promise

### ISSUE THREAD

**Issue by martintreurnicht** [wix/Detox#2579]
### Description
We have a Promise that waits for confirmation result from a dialog. This use to work fine with v17, but on v18 synchronization is waiting for all promises to complete, causing a deadlock.

I tried to disable synchronization but it seems that it isn't supported with v18
#### Reproduction

Provide the steps necessary to reproduce the issue. If you are seeing a regression, try to provide the last known version where the issue did not reproduce.

1. Create a promise that only resolves once onClick of button or view has been tapped

#### Expected behavior
promises on main thread should not wait for synchronization (as with v17) or we need to have a way to disable synchronization for main thread.

#### Environment (please complete the following information):
 - Detox: 10.1.0
 - React Native: 0.63.3
 - Node: v15.4.0
 - Device: iPhone 8 Plus
 - Xcode: 12.3
 - iOS: 14.2
 - macOS: 11.1

### Logs
```shell
detox[11736] INFO:  [actions.js] The system is busy with the following tasks:

Dispatch Queue
⏱ Queue: “Main Queue (<OS_dispatch_queue_main: com.apple.main-thread>)” with 1 work blocks

Run Loop
⏱ “Main Run Loop”

UI Elements
⏱ 0 layer animations pending
detox[11736] INFO:  [actions.js] The system is busy with the following tasks:

Dispatch Queue
⏱ Queue: “Main Queue (<OS_dispatch_queue_main: com.apple.main-thread>)” with 1 work blocks

Run Loop
⏱ “Main Run Loop”

UI Elements
⏱ 0 layer animations pending
detox[11736] INFO:  [actions.js] The system is busy with the following tasks:

Dispatch Queue
⏱ Queue: “Main Queue (<OS_dispatch_queue_main: com.apple.main-thread>)” with 1 work blocks

Run Loop
⏱ “Main Run Loop”

UI Elements
⏱ 0 layer animations pending
detox[11736] INFO:  [actions.js] The system is busy with the following tasks:

Dispatch Queue
⏱ Queue: “Main Queue (<OS_dispatch_queue_main: com.apple.main-thread>)” with 1 work blocks

Run Loop
⏱ “Main Run Loop”

UI Elements
⏱ 0 layer animations pending
detox[11736] DEBUG: [exec.js/EXEC_SUCCESS, #16] Matched Finger!
detox[11736] INFO:  [actions.js] The system is busy with the following tasks:

Dispatch Queue
⏱ Queue: “Main Queue (<OS_dispatch_queue_main: com.apple.main-thread>)” with 1 work blocks

Run Loop
⏱ “JS Run Loop”
⏱ “Main Run Loop”

One Time Events
⏱ “Runloop Perform Block” with object: “JS Run Loop”
⏱ “Runloop Perform Block” with object: “JS Run Loop”
⏱ “Runloop Perform Block” with object: “JS Run Loop”
⏱ “Runloop Perform Block” with object: “JS Run Loop”
⏱ “Runloop Perform Block” with object: “JS Run Loop”
⏱ “Runloop Perform Block” with object: “JS Run Loop”
```
#### If you are experiencing a timeout in your test
- [x] I have followed the instructions under [Lower-level Idling Resources Debug](https://github.com/wix/Detox/blob/master/docs/Troubleshooting.Synchronization.md#lower-level-idling-resources-debug-ios-only) and am providing the log below:

#### Device and verbose Detox logs
- [x] I have run my tests using the `--loglevel trace` argument and am providing the verbose log below:

**Comment by LeoNatan** [comment]
“ ⏱ 0 layer animations pending”

While the “0” is incorrectly reported, and has already been fixed in master, it looks like you have an infinite animation, likely from a spinning activity indicator view. 

You should either mock or disable infinite animations for tests.

**Comment by martintreurnicht** [comment]
@LeoNatan Not the case, this was my first thought too but i've confirmed that it's caused by a promise, like I said, this works just fine on v17, and continues to work fine on android. Once i switched from promise to callback, things started to work again.

The issue seems to be that the promise is awaiting to continue, but the `resolve()` function from the promise is triggered by the confirm/cancel buttons from the dialog buttons. This seems to create a deadlock where detox can never click the button to resolve the promise, because it's waiting for the promise to be resolved first.

**Comment by LeoNatan** [comment]
Promise is a JS concept and has nothing to do with Detox, which operates at the native layer.

**Comment by LeoNatan** [comment]
On the other hand, the idle status you provided in the log doesn’t lie. There is an infinite animation somewhere.

**Comment by martintreurnicht** [comment]
@LeoNatan Ok thanks, i will continue to investigate. In the meantime, is there some way to disableSynchronization with v18?

**Comment by LeoNatan** [comment]
Yes, same API as before.

**Comment by martintreurnicht** [comment]
@LeoNatan Hmm... Interesting, it doesn't seem to work for me. Let me try again. Thanks for the feedback

**Comment by LeoNatan** [comment]
You may need to launch the app with sync disabled, if this is on launch.

**Comment by martintreurnicht** [comment]
@LeoNatan This issue doesn't occur on launch, only happens at the very end of the test. Just for reference, 99% of our tests are working just fine, we even use the same modal in all of them, the only difference is this promise that we're using in this specific scenario that causes a deadlock, like i said we've been using v17 for many months now without any issues, issue popped up on iOS only as soon as we switched to v18. Let me investigate further, worst case scenario we'll have to replace the promise with callback, which is janky as hell, but can be done

**Comment by LeoNatan** [comment]
As I said, Detox has no awareness of your JS code, such as promises. If the log says later animation, that's what it is.

**Comment by martintreurnicht** [comment]
@LeoNatan Thanks for pushing me in the right direction, after investigating deep i realized that we show a spinner on the button when submitting, I couldn't see it because it was rendered underneath the modal. Switching this animation off for Detox builds have resolved the issue. Thanks for the help!

**Comment by lukewlms** [comment]
Just documenting here - my problem causing "The system is busy with the following tasks" was not awaiting `device.launchApp()` in the tests file.  Fixed like this:

```
 beforeAll(async() => {
  await device.launchApp()
 })
```

**Comment by jkhusanov** [comment]
Also, for anyone that could not find animations, make sure things like API calls are not in an infinite loop.

**Comment by d4vidi** [comment]
@jkhusanov FYI when running with Detox's `debug-synchronization` feature enabled (on by default), you should be able to get specific logs telling you what's holding the app back from going idle - so as to save your time on inspecting whether these are animations or network calls that are behind it.

**Comment by annuneetu** [comment]
@lukewlms I am already using  
beforeAll(async() => {
  await device.launchApp()
 })

but still facing the issue defined by you on iOS, the scripts very well on android, any suggestions how can I rectify it ?

**Comment by annuneetu** [comment]
@martintreurnicht could you please advise how you did "after investigating deeper i realized that we show a spinner on the button when submitting", as I am also suspecting the same in the AUT

**Comment by martintreurnicht** [comment]
> @martintreurnicht could you please advise how you did "after investigating deeper i realized that we show a spinner on the button when submitting", as I am also suspecting the same in the AUT

@annuneetu Nothing special. I just started removing components from the screen until it started working. In this case i actually just made the code do everything other than opening the dialog, and that is when i saw the spinner and put 2 and 2 together, just be systematic and try and narrow the scope of the issue. If this is a new implementation i would also not rule out the possibility that the promise isn't being resolved correctly
